src/Lean/Meta/Tactic/AC/Main.lean

```diff
@@ -0,0 +1,337 @@
+      | maybeNormalized opId _ _, _ => opId
+      | _, unnormalized opId _ => opId
+      | _, maybeNormalized opId _ _ => opId
+
+    let ctx ← get
+    let arbitrary := ctx.exprIdsReverse[vars[0]]
+    let preContext := ctx.preContextsReverse[op]
+    let vars : List Lean.Expr ← vars.toList.mapM fun x => do
+      let xExpr := ctx.exprIdsReverse[x]
+      let isNeutral ←
+        match ←neutralCache op x with
+        | none => mkAppOptM ``Option.none #[←mkAppM ``IsNeutral #[preContext.op, xExpr]]
+        | some isNeutral => mkAppM ``Option.some #[isNeutral]
+
+      mkAppM ``Variable.mk #[xExpr, isNeutral]
+
+    let vars ← Lean.Meta.mkListLit (←mkAppM ``Variable #[preContext.op]) vars
+
+    let comm ←
+      match preContext.comm with
+      | none => mkAppOptM ``Option.none #[←mkAppM ``IsCommutative #[preContext.op]]
+      | some comm => mkAppM ``Option.some #[comm]
+
+    let idem ←
+      match preContext.idem with
+      | none => mkAppOptM ``Option.none #[←mkAppM ``IsIdempotent #[preContext.op]]
+      | some idem => mkAppM ``Option.some #[idem]
+    return (preContext, ←mkAppM ``Lean.Data.AC.Context.mk #[preContext.op, preContext.assoc, comm, idem, vars, arbitrary])
+
+  convert (varMap : Nat → Nat) : NormalizedExpr → MetaM Lean.Expr
+    | definitelyNormalized id => mkAppM ``var #[Lean.mkNatLit $ varMap id]
+    | unnormalized _ e => convert varMap e
+    | maybeNormalized _ l r => do mkAppM ``op #[←convert varMap l, ←convert varMap r]
+
+  convertType (op : Lean.Expr) : NormalizedExpr → M Lean.Expr
+    | definitelyNormalized id => do return (←get).exprIdsReverse[id]
+    | unnormalized _ e => convertType op e
+    | maybeNormalized _ l r => do mkAppM' op #[←convertType op l, ←convertType op r]
+
+inductive ProofStrategy
+  | ac_rfl (lhs rhs : NormalizedExpr)
+  | rfl (tgt : Expr)
+  | norm (e : NormalizedExpr)
+
+def pickStrategy (e : Expr) : M ProofStrategy := do
+  match e with
+  | eq l r =>
+    match ←findUnnormalizedOperator l with
+    | lhs@(unnormalized _ _) => return ProofStrategy.norm lhs
+    | lhs@(maybeNormalized _ _ _) =>
+      match ←findUnnormalizedOperator r with
+      | rhs@(unnormalized _ _) => return ProofStrategy.norm rhs
+      | rhs => return ProofStrategy.ac_rfl lhs rhs
+    | lhs@(definitelyNormalized _) =>
+      match ←findUnnormalizedOperator r with
+      | rhs@(unnormalized _ _) => return ProofStrategy.norm rhs
+      | rhs@(maybeNormalized _ _ _) => return ProofStrategy.ac_rfl lhs rhs
+      | rhs@(definitelyNormalized _) => return ProofStrategy.rfl l
+  | e => return ProofStrategy.norm $ ←findUnnormalizedOperator e
+
+def addAcEq (mvarId : MVarId) (e : NormalizedExpr) (target : Expr) : M MVarId := do
+  let (proof, ty) ← buildProof e (←e.norm)
+  let goal ← withLocalDeclD `h_ac ty fun h_ac =>
+    mkForallFVars #[h_ac] target
+  let goal ← mkFreshExprMVar goal
+  assignExprMVar mvarId (mkApp goal proof)
+  return goal.mvarId!
+
+partial def rewriteUnnormalized (mvarId : MVarId) : M Unit :=
```

Comment:
We should implement the tactic on top of `simp`.

Reply:
done